File: lingting-spring-core/src/main/java/live/lingting/spring/util/SpringUtils.java
package live.lingting.spring.util;

import jakarta.annotation.Resource;
import live.lingting.framework.domain.ClassField;
import live.lingting.framework.util.ClassUtils;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.UtilityClass;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationContext;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * @author lingting 2022/10/15 15:21
 */
@UtilityClass
@SuppressWarnings("unchecked")
public class SpringUtils {

	@Setter
	@Getter
	private static ApplicationContext context;

	public static <T> T getBean(String name) {
		return (T) context.getBean(name);
	}

	public static <T> T getBean(Class<T> clazz) {
		return context.getBean(clazz);
	}

	public static <T> T getBean(String name, Class<T> clazz) {
		return context.getBean(name, clazz);
	}

	public static <T> Map<String, T> getBeansOfType(Class<T> type) {
		return context.getBeansOfType(type);
	}

	public static String[] getBeanNamesForType(Class<?> type) {
		return context.getBeanNamesForType(type);
	}

	public static <T> T ofBean(Class<T> cls)
			throws InvocationTargetException, InstantiationException, IllegalAccessException {
		Constructor<T>[] constructors = ClassUtils.constructors(cls);
		return ofBean(constructors[0], SpringUtils::getBean);
	}

	public static <T> T ofBean(Constructor<T> constructor, Function<Class<?>, Object> getArgument)
			throws InvocationTargetException, InstantiationException, IllegalAccessException {
		Class<?>[] types = constructor.getParameterTypes();
		List<Object> arguments = new ArrayList<>();

		for (Class<?> cls : types) {
			Object argument = getArgument.apply(cls);
			arguments.add(argument);
		}

		T t = constructor.newInstance(arguments.toArray());
		// 自动注入
		ClassField[] cfs = ClassUtils.classFields(t.getClass());
		for (ClassField cf : cfs) {
			if (!cf.hasField() || cf.isFinalField()) {
				continue;
			}
			boolean isAutowired = cf.getAnnotation(Autowired.class) != null || cf.getAnnotation(Resource.class) != null;
			if (isAutowired) {
				Class<?> cls = cf.getValueType();
				Object arg = getArgument.apply(cls);
				cf.set(t, arg);
			}
		}
		return t;
	}

}

File: lingting-spring-core/src/main/kotlin/live/lingting/spring/util/SpringUtils.kt
package live.lingting.spring.util

import jakarta.annotation.Resource
import live.lingting.framework.util.ClassUtils
import live.lingting.framework.util.ClassUtils.constructors
import org.springframework.beans.factory.NoSuchBeanDefinitionException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationContext
import java.lang.reflect.Constructor
import java.util.function.Function
import kotlin.reflect.KClass

/**
 * @author lingting 2022/10/15 15:21
 */
@Suppress("UNCHECKED_CAST")
object SpringUtils {

    @JvmField
    val AUTOWIRE_ANNOTATIONS = setOf(
        Autowired::class.java.name,
        Resource::class.java.name,
        "javax.annotation.Resource"
    )

    var context: ApplicationContext? = null

    @JvmStatic
    fun hasBean(name: String): Boolean {
        return context!!.containsBean(name)
    }

    @JvmStatic
    fun hasBean(cls: Class<*>): Boolean {
        try {
            val map = getBeansOfType(cls)
            return map.isNotEmpty()
        } catch (_: NoSuchBeanDefinitionException) {
            return false
        }
    }

    fun hasBean(cls: KClass<*>) = hasBean(cls.java)

    @JvmStatic
    fun <T> getBean(name: String): T {
        return context!!.getBean(name) as T
    }

    @JvmStatic
    fun <T> getBean(clazz: Class<T>): T {
        return context!!.getBean<T>(clazz)
    }

    fun <T : Any> getBean(cls: KClass<T>): T = getBean(cls.java)

    @JvmStatic
    fun <T> getBean(name: String, clazz: Class<T>): T {
        return context!!.getBean<T>(name, clazz)
    }

    fun <T : Any> getBean(name: String, cls: KClass<T>): T = getBean(name, cls.java)

    @JvmStatic
    fun <T> getBeansOfType(type: Class<T>): Map<String, T> {
        return context!!.getBeansOfType<T>(type)
    }

    fun <T : Any> getBeansOfType(cls: KClass<T>): Map<String, T> = getBeansOfType(cls.java)

    @JvmStatic
    fun getBeanNamesForType(type: Class<*>): Array<String> {
        return context!!.getBeanNamesForType(type)
    }

    fun getBeanNamesForType(cls: KClass<*>): Array<String> = getBeanNamesForType(cls.java)

    @JvmStatic
    fun <T> ofBean(cls: Class<T>): T {
        val constructors: Array<Constructor<T>> = constructors<T>(cls)
        return ofBean<T>(constructors[0])
    }

    @JvmStatic
    fun <T> ofBean(cls: Class<T>, getArgument: Function<Class<*>, Any?>): T {
        val constructors: Array<Constructor<T>> = constructors<T>(cls)
        return ofBean<T>(constructors[0], getArgument)
    }

    fun <T : Any> ofBean(cls: KClass<T>): T = ofBean(cls.java)

    @JvmStatic
    fun <T : Any> ofBean(cls: KClass<T>, getArgument: Function<Class<*>, Any?>): T {
        return ofBean(cls.java, getArgument)
    }

    @JvmStatic
    fun <T> ofBean(constructor: Constructor<T>): T {
        return ofBean(constructor) { getBean(it) }
    }

    @JvmStatic
    fun <T> ofBean(constructor: Constructor<T>, getArgument: Function<Class<*>, Any?>): T {
        return ClassUtils.newInstance(constructor, getArg = getArgument)
    }

}
